Update install instructions for final BuildStream 2.0 release

ssssam opened the pull request:
See commits for rationale for the various changes.

nanonyme commented on doc/source/main_install.rst:
```diff
@@ -285,13 +327,23 @@ A regular way to do this is to add the following line to the end of your ``~/.ba
+Note that some special configuration is often needed to run BuildStream in a container:
+
+  * User namespaces are used to isolate and control builds. This requires the
+    Docker ``--privileged`` mode.
```
With podman at least this is not required. We (freedesktop-sdk) have an image that works *just fine* in unprivileged toolbox under user namespaces.

nanonyme replied:
Right, so --privileged I think means different things here. With Docker it means "you can privilege escalate to root", with podman it means "you can privilege escalate to user".

nanonyme replied:
The whole point of user namespaces is they work without privileges.

nanonyme replied:
You can check with `registry.gitlab.com/freedesktop-sdk/infrastructure/freedesktop-sdk-docker-images/bst2:latest`

nanonyme replied:
Fair. So the documentation is correct, it just might just be in podman case be less privileged than docker.

abderrahim replied:
@nanonyme AFAIK toolbx uses a `--privileged` podman container. You're probably confusing rootless with non-privileged.

gtristan commented on doc/source/main_install.rst:
```diff
@@ -130,28 +116,23 @@ Installing BuildStream
-This will install latest dev snapshot of BuildStream and its pure python
-dependencies into your user's homedir in ``~/.local``.
+This will install BuildStream and its dependencies into your user's homedir in
+``~/.local``.  Pip will use binary "wheel" packages from PyPI where these are
```
Nit:

> Pip will use binary "wheel" packages from PyPI where these are available for your platform.

Would probably read better with *"if these are available for your platform"* (whereas *"where"* would be more suitable if *"for your platform"* was removed).

gtristan commented:
Overall, I like the install from source instructions here, however it looks like this is unfortunately again growing into a lot of overlap with the user facing instructions at: https://buildstream.build/install.html

Lets keep the user facing install paths published at https://buildstream.build/install.html and have the version specific developer/packager facing instructions at https://docs.buildstream.build/master/main_install.html. This makes the two pages serve obviously different purposes, and makes the reading of either of these less crowded and intense.

Maybe it would be helpful to have a little map:

```
                            Main page (A)
                             /       \
                           /           \
            Source (B-1), (B-2)       Docker (C)

(A): https://buildstream.build/install.html
(B-1): https://docs.buildstream.build/master/main_install.html
(B-2): https://docs.buildstream.build/1.6/main_install.html
(C): https://gitlab.com/BuildStream/buildstream-docker-images/-/blob/master/USING.md
```

While sub-pages might link back to the main user facing install page on the main website (A), we should always assume (A) as a starting point, and (A) should always be our answer for anyone's question of *"How do I install BuildStream"*

### Introduction

I think the statement:

> BuildStream is packaged in some Linux distributions. If your distro has an up-to-date package we recommend using that.

Can go away, this is already mentioned where it is better suited, at: https://buildstream.build/install.html

It could be good to have a forward in `main_install.rst` saying something like:

> This page explains how to build and install this version of BuildStream from sources. If you are trying to install BuildStream on your computer it is recommended to try following the [general purpose installation instructions here](https://buildstream.build/install.html) first, and return here only if you need to build and install from source.

### Installation from PyPI

I think this can all go away and is not relevant here, and is already explained where it is better suited at: https://buildstream.build/install.html

### Docker images

I think the whole *"Using docker images"* section here can also go away completely.

If there are user facing images ready to be used by users, we should trust https://gitlab.com/BuildStream/buildstream-docker-images/-/blob/master/USING.md to document how to use the latest BuildStream 1 and BuildStream 2 images respectively, and link to that page from the user facing documentation at https://buildstream.build/install.html (without going into a whole detailed shpiel about it there either)...

And of course we should touch up https://gitlab.com/BuildStream/buildstream-docker-images/-/blob/master/USING.md a bit to make sure it at least tells you the registry and image names which are useful to use, and posts some useful invocation like:

```shell
docker run -priviliged -it -v `pwd`:/build <bstimagename:latest> /bin/bash
```

### Build/run dependencies

When discussing installing dependencies, here we are talking about the dependencies needed to install *from source*, so it is unnecessary to split up dependencies as we currently have them down (redundantly on a per-distro basis on top of it), with statements like: *"Run this command"* and then if that is not enough, also *"Run this other command too"*.

Also, this would look so much more concise if it were simply a condensed table with two columns:

| Distribution        | Command           |
| ------------- |-------------|
| Arch      | sudo pacman -S bubblewrap fuse3 git lzip patch python python-pip gcc |
| Debian  | sudo apt-get install git lzip patch bubblewrap fuse3 python3-pip python3 g++ python3-dev |


Otherwise right now, it looks like you have to scroll through pages of documentation just to install the distro dependencies needed to build and install BuildStream.

### Checking out git

When talking about checking out git (which are the only installs we need to talk about on this page I guess), it is unnecessary to talk about *Upgrading* separately.

If we have instructed the user how to checkout a useful tag/branch which contains what they want, then they already have the ammunition they need go ahead and install a different or newer version later.

### Install instructions

When discussing the actual installation instructions, it is worth noting:

* `pip3 install --user .` as a default just to conveniently install
* `python3 setup.py install` it is important to note that this is supported because this is the regular distro packager path, which does not involve downloading dependencies from the non-existing internet

We should also have a simple reminder here that depending on where you have decided to install buildstream, you may need to adjust `$PATH` or *activate* your virtual environment in order to use it.

This last part will allow us to nuke that whole **Post-install setup** section which merely talks about `$PATH`

#### Virtual environments

It is also worth noting as part of install instructions that you can use a python virtual environment if you want to have multiple versions of BuildStream installed on the same machine.

An interesting question is whether to duplicate virtual environment instructions on the main page (using PyPI packages) and here on the source installation page (installing from git).

If we continue to link to this page from the main page in order to discuss virtual environments, we should use an encoded link anchor in the reStructuredText for this:

```
.. _install-virtual-environment:
```

Because the main website has a link to: https://docs.buildstream.build/master/main_install.html#installing-in-virtual-environments which only happens to work because we didn't yet haphazardly rename the title of a section in our docs.

### Completions

We should restore the section about *completions* as this is an important detail for distro packagers and also users.

For this we can merely link: https://github.com/apache/buildstream/blob/master/src/buildstream/data/bst

We can safely say (like the file says) where it should be installed, and we can even just paste the contents of that script and inform users that they can just append that text to `~/.bash_completion` (this completion script never changes so it's quite safe).

---

After trimming out all that fat above, this would leave us with a more manageable structure of only:

* Introduction
* Dependencies for building/running this version of buildstream
* Installing buildbox
  * Instructions about binaries
  * Link(s) to how to build buildbox from sources in upstream buildbox
* Cloning the right git tag/branch
* Installation instructions (from the checkout)
  * Virtual environments
* Completions

ghost commented:
I'm taking another look at this today, thanks for the detailed guidance.

On dependencies and package lists, I will move them into a table, but we should still separate the install-time requirements (gcc, pip, python3-devel) from the runtime requirements. Let's try a table with 2 columns.

About a suitable 'install' command for distro packagers, it's useful to give some guidance on this. The `python3 setup.py install` command will work (as long as Bst doesn't change buildsystem), but it may not be what distro package guidelines recommend now. did some quick research to satisfy my own curiousity:

  * Fedora recommends use of [a %pyproject_install macro](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#_build_macros), which runs `pip install --no-deps --no-index --use-pep517 .` ([link](https://src.fedoraproject.org/rpms/pyproject-rpm-macros/blob/rawhide/f/macros.pyproject#_60))
  * Debian doesn't give any strong recommendation, the existing [buildstream package](https://tracker.debian.org/pkg/buildstream) uses [pybuild](https://manpages.debian.org/bullseye/dh-python/pybuild.1.en.html) via debhelper, which autodetects one of multiple backends. Running `pybuild --detect` in buildstream.git reports `distutils`, so they do indeed [run `setup.py` directly](https://salsa.debian.org/python-team/tools/dh-python/-/blob/master/dhpython/build/plugin_distutils.py), although there is a `pyproject` backend that may become default in future which [runs `python3 -m build`](https://salsa.debian.org/python-team/tools/dh-python/-/blob/master/dhpython/build/plugin_pyproject.py) and post-processes the generated `.whl`.

All this to highlight there's no "one size fits all" distro packager command, they should use whatever the Python guidelines for that distro recommends, and that can be either running setuptools directly or any PEP517 compatible frontend.

Commit:
doc/source/main_install.rst: Rework as "Install from source" guide

For rationale, see:
https://github.com/apache/buildstream/pull/1741#issuecomment-1227820050

Changes include:

  * provide a list of host tools required by source plugins and by
    BuildBox - almost no real-world projects can be built unless these
    are installed.
  * clarify which dependencies are needed only at install time
  * recommend use of latest BuildBox tag, not specific outdated tags
  * remove "install from PyPI" instructions - this belongs on the
    website
  * remove container install docs - these belong on website
  * add info about Bash completions
  * BuildStream 2 is now "in beta", this sounds more complete than "dev
    snapshot".
  * avoid "simply", nothing in tech is ever simple :)

Commit:
doc/source/core_plugins.rst: List buildstream-plugins

ghost commented:
Updated following the above guidance.

gtristan commented on doc/source/main_install.rst:
```diff
@@ -1,221 +1,182 @@
-This will install BuildStream's pure python dependencies into
-your user's homedir in ``~/.local`` and will run BuildStream directly
-from the git checkout directory.
+You can optionally use Pip's
```
I don't believe the `-e` option has worked for a very long time (with buildstream specifically, this might have been since the transition to use cython).

gtristan commented on doc/source/main_install.rst:
```diff
@@ -258,40 +219,23 @@ and set up other necessary environment variables::
 
-   You will have to restart your terminal in order for these changes to take effect.
+BuildStream integrates with Bash to provide helpful tab-completion. These
+are provided by the script ``src/buildstream/data/bst`` which requires
```
Wouldn't it be good to expand the content of `src/buildstream/data/bst` here (with `::literalinclude`) or at least provide a link to the master branch version of the `bst` completions script on github ?

Also I thought there was an easier way to get this working for a single user, by simply copying the content of the script (from this page) and pasting the content directly into `~/.bash_completion`.

Commit:
doc/source/main_install.rst: Dis-recommend `pip install -e`

Commit:
doc/source/main_install.rst: Rework completions instructions

ghost commented:
Last 2 review comments now addressed in 7925769d9

gtristan committed:
main_install.rst: specify dependencies from requirements.in

The requirements.txt file specifies the frozen dependencies used in CI,
the requirements.in are the actual requirements checked by setup.py
